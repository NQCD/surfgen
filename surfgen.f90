
!******************************************************************************
! MAIN PROGRAM
program surfgen
    use progdata
    implicit none
    integer   ::jobtype

    print *,"Entering surfgen.global"
    call readinput(jobtype)
    call initialize(jobtype)

    select case(jobtype)
     case(1)
      if(printlvl>0)print *,"Reading displacement data..."
      call readdisps()
      if(printlvl>0)print *,"Making surface..."
      call makesurf()
     case(2)
      if(printlvl>0)print *,"Performing critical point searches..."
      call minmex()
     case(3)
      if(printlvl>0)print *,"Loading geometries..."
      call loadgeom()
    end select

  call cleanup()

end program surfgen
! read general input from input file (surfgen.in)
! and read in job specific input according to jtype
SUBROUTINE readinput(jtype)
  use progdata
  use hddata, only: initGrps,ncoord,nstates,order
  use CNPI, only: irrep,grpPrty,grpSym
  IMPLICIT NONE
  INTEGER,INTENT(INOUT)           :: jtype
  INTEGER                         :: i,j,mkadiabat,ios
  INTEGER                         :: nGrp,jobtype
  INTEGER,dimension(10)           :: surface,updatehess
  INTEGER,dimension(50)           :: ci,atmgrp
  DOUBLE PRECISION,dimension(500) :: minguess,mexguess
  INTEGER,DIMENSION(10)           :: groupSym,groupPrty
  DOUBLE PRECISION,dimension(20)  :: e_guess

  NAMELIST /GENERAL/        jobtype,natoms,order,nGrp,groupsym,groupprty,usefij,switchdiab,&
                            printlvl,deg_cap,inputfl,eshift,atmgrp,use_eguess,e_guess
  NAMELIST /MINMEX/         nmin,nmex,minguess,mexguess,optiter,opttoler,updatehess,h_recal,surface,&
                             ci,maxstep,xscale,sscale,degtoler,enforcepd
  NAMELIST /LOADGEOM/       isloop,loopst,ngeoms,geomfl,calchess,outputdir,outputdiab
 
  jtype      = 0
  natoms     = 2
  printlvl   = 1
  inputfl    = ''
  usefij     = .true.
  deg_cap    = 1D-7
  eshift     = dble(0)
  use_eguess = .false.
  switchdiab = .false.
  print *,"Entering readinput()."
 !----------- GENERAL SECTION ----------------!
  open(unit=INPUTFILE,file='surfgen.in',access='sequential',form='formatted',&
       IOSTAT=ios,POSITION='REWIND',ACTION='READ',STATUS='OLD')
  if(ios/=0)then
    print *,"readinput: cannot open file surfgen.in.  IOSTAT=",ios
  end if!ios/=0

  read(unit=INPUTFILE,NML=GENERAL)
  if(printlvl>0)print *,"    readinput():  Control parameters read from surfgen.in"
  jtype  = jobtype

  call genAtomList(atmgrp)
  
  call readIrreps()
  if(allocated(GrpSym))deallocate(GrpSym)
  if(allocated(grpPrty))deallocate(grpPrty)
  allocate(GrpSym(nGrp))
  allocate(grpPrty(nGrp))
  GrpSym=groupsym(1:nGrp)
  grpPrty=groupprty(1:nGrp)
  call initGrps(nGrp,irrep(GrpSym(:))%Dim)

  if(allocated(eguess))deallocate(eguess)
  allocate(eguess(nGrp))
  if(use_eguess)then
    eguess=e_guess(1:nGrp)/AU2CM1
    print *,"Guess energies: ", eguess
  else
    do i=1,nGrp
     eguess(i)=i*10000/AU2CM1
    end do
  end if

  if(printlvl>0)print *,"    Reading job specific inputs."
  !------------ JOBTYPE SECTION ------------!

  select case(jtype)

   !-----------------------------------------------------
   ! Potlib interface.  Do nothing.
   !----------------------------------------------------
   case(0)
    if(printlvl>0)print *,"    readinput():  jobtype=0.  Ready for evaluation."

   !-----------------------------------------------------
   ! Read in displacemnts, construct surface
   !----------------------------------------------------
   case(1)
    if(printlvl>0)print *,"    readinput():  jobtype=1.  Reading displacements"
    CALL readMAKESURF(INPUTFILE)
   !--------------------------------------------------
   ! Find extrema on the surface
   !--------------------------------------------------
   case(2)
    maxstep=0.2
    minguess=dble(0)
    mexguess=dble(0)
    updatehess=0
    surface=0
    ci=0
    minstates=0
    mexstates=0
    nmin = 0
    nmex = 0
    optiter = 3
    opttoler = 1D-3
    degtoler = 1D-2
    xscale   = 1D0
    sscale   = 1D0
    h_update(1) = .false.
    h_recal = 1
    enforcepd = .false.

    read(unit=INPUTFILE,NML=MINMEX)
    if(nmin.gt.0)allocate(minstart(3*natoms,nmin))
    if(nmex.gt.0)allocate(mexstart(3*natoms,nmex))

    do i = 1,nmin
     if(surface(i).gt.0.and.surface(i).le.nstates)minstates(i) = surface(i)
     if(updatehess(i).eq.0)then
      h_update(i) = .false.
     else
      h_update(i) = .true.
     endif
     do j = 1,3*natoms
      minstart(j,i) = minguess((i-1)*3*natoms+j)
     enddo
    enddo

    do i = 1,nmex
     mexstates(i,:) = ci(i*2-1:i*2)
     do j = 1,3*natoms
      mexstart(j,i) = mexguess((i-1)*3*natoms+j)
     enddo
    enddo

   !-------------------------------------------------------
   ! Load geometries, print out geom/frequency information
   !-------------------------------------------------------
   case(3)
    isloop   = .false.
    loopst   = 1
    ngeoms   = 1
    geomfl   = ''
    calchess = .true.
    outputdiab=.false.
    outputdir= ''
    read(unit=INPUTFILE,NML=LOADGEOM)
    if(ngeoms<1)ngeoms=1
    if(.not.isloop)then
      loopst=1
      ngeoms =1
    end if
  end select

  close(unit=INPUTFILE)


  if(printlvl>0)print *,"Exiting readinput..."
  return
end SUBROUTINE readinput
