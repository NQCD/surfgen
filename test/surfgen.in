&COMMENTS
jobtypes: 
1 = read reference, make surface, write surface to disk
2 = read reference, read surface, find extrema
3 = read reference, read surface, reshape surface
4 = read reference, read surface, output surface in new coordinate system
5 = read reference, load geometry information

Irreducible representation matrices are given in irrep.in
states are partitioned into groups that each n-state group carries an n-dimensional irrep.  
groupsym specifies the irreducible those groups carry.
The numbers correspond to the index in irrep.in.
groupprty specifies the parity of each group.

/
&GENERAL
printlvl   = 3,
jobtype    = 1, 
natoms     = 4,
atmgrp     = 1,2,2,2
order      = 3,
ngrp       = 2,
groupsym   = 1,2
groupprty  = 1,-1
inputfl    = 'hd.data'
/
&MAKESURF
eshift     = 56.475986184916D0 
usefij     = .true.
deg_cap    = 1D-5
dijscale   = 1d0
energyT    = 5.5D4,6D4,7.0D4,1D5,1D6,1D7
highEScale = 5D-1,2D-1,5D-2,1D-2,1D-3,1D-4
gscalemode = 0
useIntGrad = .true.
linsteps   = 0
linNegSteps= 0
stepmethod = 0
maxd       = 1D1
intGradT   = 1D-2
intGradS   = 1D-1
TBas       = 1D-13
rmsexcl    = -2
npoints    = 2553 
dfstart    = 0
gcutoff    = 1D-7
maxiter    = 20  
toler      = 1D-7
LSETol     = 1.0D-7
flattening = 1.0D-8 
exactTol   = 1.0D-9 
gorder     = 5000.
outputfl   = 'hd.data'
flheader   = 'hd coefficient file of nh3 made with surfgen'
w_energy   = 1D0
w_grad     = 2D-1
w_fij      = 2D-1
nrmediff   = 1D3
ediffcutoff= 1D2
nrmediff2  = 2d2
ediffcutoff2=1d2   
ndstart    = 200
ndiis      = 35 
/
&MINMEX
nmin       = 2,
nmex       = 0,
minguess   = 0.22,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.
            -0.22,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.
mexguess   = 0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,
optiter    = 1000,
opttoler   = 1E-8,
surface    = 1,1
ci         = 1,2,0, 1,2,3
usehessian = 1,1,
hessupdate = 25,
/
&RESHAPE
nmin       = 2,
nept       = 0,
cmethod    = 1,
fixiter    = 25,
conlist    = 3,1,1, 3,2,2, 3,3,3, 3,4,4, 3,5,5, 3,6,6, 3,7,7, 3,8,8, 3,9,9, 3,10,10, 3,11,11, 3,12,12, 3,13,13, 3,14,14, 3,15,15,
             3,16,16, 3,17,17, 3,18,18
minguess   = 0.22,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.
            -0.22,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.,0.
optiter    = 2000,
opttoler   = 1E-7,
surface    = 1,1
usehessian = 1,1
hessupdate = 25,
makemap    = 1,1
modeweight = 1.,1.,1.,1.,1.,0.05,0.05,1.,1.,1.,1.,1.,1.,1.,1.,1.,1.,0.05,
             1.,1.,1.,1.,1.,0.05,0.05,1.,1.,1.,1.,1.,1.,1.,1.,1.,1.,0.05,
funcweight = 0.01,0.05,0.89,0.05,0.0,
/
&TRANSFORM
mkadiabat  = 1,
nbasis     = 2,
/
&LOADGEOM
isloop     = .true.
loopst     = 1
ngeoms     = 20
geomfl     = 'geom'
calchess   = .false.
outputdir  = 'path4'
/
