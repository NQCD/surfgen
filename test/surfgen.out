-----------------------------------------------------------
                                                           
                     surfgen.global                        
                                                           
   * Creates quasi-diabatic Hamiltonian by reproducing     
   ab initio energy gradients and derivative couplings     
   with selected data fitted exactly and the rest in a     
   least-squares sense                                     
                                                           
   * Blocks of Hd are expanded as polynomials of scaled    
   internuclear distance and out-of-plane angle coords     
   that are well-defined globally.                         
                                                           
   * Hd is constructed, using projection operators, so     
   that the diabats carry certain irreps of CNPI group     
                                                           
   Xiaolei Zhu,  2010                                      
   Department of Chemistry, Johns Hopkins University       
   based on SURFGEN.X, Michael Schuurman,  2008            
-----------------------------------------------------------

  jobtype = POTLIB   


